src/ast/recfun_decl_plugin.cpp

```diff
@@ -0,0 +1,434 @@
+        compute_case_state(ast_manager & m)
+            : m_manager(m),
+              m_region(),
+              m_case_idx(0),
+              m_subst(m, false, false),
+              m_ssubst(m_subst),
+              m_ops(),
+              m_ops_idx(0),
+              m_alternatives()
+            {}
+
+        ast_manager &           m() { return m_manager; }
+        region &                reg() { return m_region; }
+        expr_substitution &     subst() { return m_subst; }
+    };
+
+    // an action (explore term, etc.) in `compute_cases`
+    struct compute_case_op {
+        virtual void operator()(compute_case_state & st) = 0;
+        virtual void undo(compute_case_state & st) = 0;
+    };
+
+    // to explore alternative branches in `compute_cases`
+    class compute_case_alternative {
+        unsigned cond_size; // m_conditions.size()
+        unsigned ops_idx; // m_ops_idx
+        unsigned ops_size; // m_ops.size()
+    public:
+        compute_case_alternative(compute_case_state & st)
+            : cond_size(st.m_conditions.size()),
+              ops_idx(st.m_ops_idx),
+              ops_size(st.m_ops.size())
+              {}
+
+        // called when backtracking; can push stuff on `ops` stack
+        virtual void operator()(compute_case_state & st) {
+            st.m_ops_idx = ops_idx;
+            st.m_ops.shrink(ops_size);
+            st.m_conditions.shrink(cond_size);
+        }
+    };
+
+    // explore given term (typically, explore subterms)
+    class compute_case_explore : public compute_case_op {
+        expr * e;
+    public:
+        compute_case_explore(expr * e) : compute_case_op(), e(e) {}
+        void undo(compute_case_state & st) override {}
+        void operator()(compute_case_state & st) override;
+    };
+
+    class compute_case_ite_alt;
+
+    // explore `ite a b c` by exploring `b` (resp. `c`) under condition `a` (resp. `¬a`)
+    class compute_case_ite : public compute_case_op {
+        ast_manager & m;
+        app * e; // ite
+        bool pos_case; // explore positive case or negative case?
+    public:
+        compute_case_ite(ast_manager & m, app * e0, bool pos_case=true)
+            : compute_case_op(), m(m), e(e0), pos_case(pos_case)
+            { SASSERT(m.is_ite(expr)); }
+
+        void operator()(compute_case_state & st) override;
+        void undo(compute_case_state & st) override {
+            st.m_ssubst.pop(1);
+        }
+    };
+
+    void compute_case_explore::operator()(compute_case_state & st) {
+        if (st.m().is_ite(e)) {
+            app * a = static_cast<app*>(e);
+            compute_case_op * op = new (st.reg()) compute_case_ite(st.m(), a, true);
+            st.m_ops.push_back(op);
+        }
+        else if (is_app(e)) {
+            // explore all subterms
+            app * a = static_cast<app*>(e);
+            for (unsigned i=0; i < a->get_num_args(); ++i) {
+                compute_case_op * op = new (st.reg()) compute_case_explore(a->get_arg(i));
```

Comment:
It could get tricky with using regions since destructors, and therefore reference counting on objects, will be bypassed.